core/src/main/java/io/grpc/ManagedChannelBuilderProvider.java

```diff
@@ -0,0 +1,94 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ServiceLoader;
+
+/**
+ * Provider of managed channels for transport agnostic consumption.
+ *
+ * <p>Implementations <em>should not</em> throw. If they do, it may interrupt class loading. If
+ * exceptions may reasonably occur for implementation-specific reasons, implementations should
+ * generally handle the exception gracefully and return {@code false} from {@link #isAvailable()}.
+ */
+@Internal
+public abstract class ManagedChannelBuilderProvider {
+  private static final ManagedChannelBuilderProvider provider;
+
+  static {
+    ServiceLoader<ManagedChannelBuilderProvider> providers
+        = ServiceLoader.load(ManagedChannelBuilderProvider.class);
+    Iterator<ManagedChannelBuilderProvider> iter = providers.iterator();
+    List<ManagedChannelBuilderProvider> list = new ArrayList<ManagedChannelBuilderProvider>();
+    for (ManagedChannelBuilderProvider current : providers) {
+      if (!current.isAvailable()) {
+        continue;
+      }
+      list.add(current);
+    }
+    if (list.isEmpty()) {
+      provider = null;
+    } else {
+      provider = Collections.max(list, new Comparator<ManagedChannelBuilderProvider>() {
+        @Override
+        public int compare(ManagedChannelBuilderProvider f1, ManagedChannelBuilderProvider f2) {
+          return f1.priority() - f2.priority();
+        }
+      });
+    }
+  }
+
+  public static ManagedChannelBuilderProvider provider() {
+    return provider;
+  }
+
+  /**
+   * Whether this provider is available for use, taking the current environment into consideration.
+   * If {@code false}, no other methods are safe to be called.
+   */
+  protected abstract boolean isAvailable();
+
+  /**
+   * A priority, from 0 to 10 that this provider should be used, taking the current environment into
+   * consideration. 5 should be considered the default, and then tweaked based on environment
+   * detection. A priority of 0 does not imply that the provider wouldn't work; just that it should
+   * be last in line.
+   */
+  protected abstract int priority();
```

Comment:
@carl-mastrangelo, I've found priority quite common in plugin scenarios. Most advanced plugin systems I've seen have had some sort of priority, to either sort initialization order or to prefer one plugin over another. It only ever seems to work "okay," but I've also not seen anything work better.

But you are right, to choose a good and proper value, you need to be aware of what other providers are providing. But we can do that are development time and not have the runtime be forced to deal with a partial ordering of options.